src/basic/fd-util.c

```diff
@@ -864,3 +864,15 @@ int fd_get_diskseq(int fd, uint64_t *ret) {
 
         return 0;
 }
+
+int dir_fd_is_root(int dir_fd) {
+        struct stat st, rst;
+
+        assert(dir_fd >= 0);
+
+        if (fstat(dir_fd, &st) < 0 || lstat("/", &rst) < 0)
```

Comment:
hmm, i'd avoid using absolute paths for this, given the complexity of pivot_root() vs chroot().

maybe instead check if the inode and its parent inode have the same mnt_id.

Would require some reworking of path_get_mnt_id() to also work with fds (which is trivial, all internal codepaths support that).

This uses the fact that ".." from the root inode will point you onto the root inode again, without leaving the mount.

Reply:
I'm confused, how does comparing mount ids of the fd and its parent tell us if the fd is the root fd?

Reply:
Nvm I'm stupid